File: 102. Binary Tree Level Order Traversal/levelOrder.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    var queue = [root]
    var result = []
    
    while (queue.length > 0) {
        let level = []
        let q_len = queue.length
        for (i=0;i<q_len;i++) {
            let node = queue.shift()
            if (node) {
                level.push(node.val)
                if (node.left) {
                    queue.push(node.left)
                }
                if (node.right){
                    queue.push(node.right)
                }                
            }
        }
        
        if (level) {
            result.push(level)
        }
    }
    if (root) {
        return result
    }
    else {
        return []
    }
};

// Runtime 80 ms / Memory 44.5 MB
// debug: https://jsfiddle.net/

File: 102. Binary Tree Level Order Traversal/levelOrder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> output;
        // empty
        if (root == NULL) { return output; }
        // BFS
        queue<TreeNode*> nodeList;
        nodeList.push(root);

        while (!nodeList.empty()) {
            // same depth            
            int n = nodeList.size();
            vector<int> valueList;
            for(int i=0; i<n; i++){
                // popup(0)
                TreeNode* node = nodeList.front();
                nodeList.pop();
                valueList.push_back(node->val);
                // push_back
                if(node->left != NULL) { nodeList.push(node->left); }
                if(node->right != NULL) { nodeList.push(node->right); }
            }
            // next depth
            output.push_back(valueList);
        }

        return output;
    }
};